File: 02_agent_patterns/01_agent_loop_basics/agent-loop-target/go-example/normalize.go
package agentloop

// Normalize scales values into the range [0, 1] by dividing by the maximum.
//
// Edge cases the caller relies on:
//   - An empty slice returns an empty slice.
//   - If the maximum is 0 (e.g. all values are 0), every result is 0.0 rather
//     than producing NaN.
func Normalize(values []float64) []float64 {
	largest := values[0]
	for _, v := range values {
		if v > largest {
			largest = v
		}
	}
	out := make([]float64, len(values))
	for i, v := range values {
		out[i] = v / largest
	}
	return out
}

File: 02_agent_patterns/01_agent_loop_basics/agent-loop-target/agent_loop_target/normalize.py
"""Bug 3 (missing edge case): scale a list of numbers to the 0..1 range."""


def normalize(values):
    """Scale ``values`` into the range [0, 1] by dividing by the maximum value.

    Edge cases the caller relies on:
      * An empty list returns ``[]``.
      * If the maximum is 0 (e.g. all values are 0), every result is ``0.0``
        rather than raising a division error.

    Args:
        values: a list of non-negative numbers.

    Returns:
        A new list of floats in [0, 1], the same length as ``values``.
    """
    largest = max(values)
    return [v / largest for v in values]
